File: dotnet/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/Handlers/PaymentValidationHandler.cs
namespace ChainOfResponsibilityDesignPattern.Example1.Handlers;

public class PaymentValidationHandler : BaseOrderHandler
{
    private readonly Dictionary<string, decimal> _customerBalances;

    public PaymentValidationHandler()
    {
        _customerBalances = new Dictionary<string, decimal>
        {
            { "John Doe", 5000m },
            { "Jane Smith", 1500m },
            { "Bob Johnson", 500m },
            { "Alice Williams", 10000m }
        };
    }

    public override void Handle(OrderRequest request)
    {
        if (!_customerBalances.ContainsKey(request.CustomerName))
        {
            request.AddMessage($"Payment failed: Customer '{request.CustomerName}' not found");
            request.IsApproved = false;
            return;
        }

        var customerBalance = _customerBalances[request.CustomerName];

        if (customerBalance >= request.TotalAmount)
        {
            _customerBalances[request.CustomerName] -= request.TotalAmount;
            request.AddMessage($"Payment validated: ${request.TotalAmount:F2} charged");
            base.Handle(request);
        }
        else
        {
            request.AddMessage($"Insufficient balance: Need ${request.TotalAmount:F2}, available ${customerBalance:F2}");
            request.IsApproved = false;
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ChainOfResponsibilityDesignPattern/ChainOfResponsibilityDesignPattern.Example1/src/Handlers/PaymentValidationHandler.ts
import { BaseOrderHandler } from "../BaseOrderHandler";
import { OrderRequest } from "../OrderRequest";

export class PaymentValidationHandler extends BaseOrderHandler {
	private readonly _balances: Map<string, number> = new Map([
		["John Doe",       5000],
		["Jane Smith",     1500],
		["Bob Johnson",     500],
		["Alice Williams", 10000],
	]);

	handle(request: OrderRequest): void {
		const balance = this._balances.get(request.customerName);
		if (balance === undefined) {
			request.addMessage(`Payment failed: Customer '${request.customerName}' not found`);
			request.isApproved = false;
			return;
		}
		if (balance >= request.totalAmount) {
			this._balances.set(request.customerName, balance - request.totalAmount);
			request.addMessage(`Payment validated: $${request.totalAmount.toFixed(2)} charged`);
			super.handle(request);
		} else {
			request.addMessage(`Insufficient balance: Need $${request.totalAmount.toFixed(2)}, available $${balance.toFixed(2)}`);
			request.isApproved = false;
		}
	}
}
